### PR TITLE
feat: Integrate FastHTML replacing FastAPI and React

### PR DESCRIPTION
**Pull request by jamilraichouni**
This introduces several major changes by

- replacing the FastAPI backend and React frontend with FastHTML and
- replacing the Makefile with a Python CLI tool that manages the project.
 
The project can be managed including

- building the Docker image,
- running the app in a container,
- running the app locally with optional full hot code reloading, and
- building the Tailwind CSS stylesheet.

New is also a hot template reloading, which is enabled per default.

In development mode, it also runs with hot code reloading.

The FastHTML integration involves reusing existing code where possible, but also includes significant rewrites that affect the application's behavior. This update aims to enhance performance and streamline the development process.

# Open issues/ topics

1. [x] fix: Make env var `ROUTE_PREFIX` work for static files
1. [x] chore: Build stylesheet always via Tailwind CLI.
1. [x] fix: CI job `docker` must set tool versions via three `--build-arg` specs
1. [x] fix: Improve CSS for rendered templates output
1. [x] fix: Make CME_ROUTE_PREFIX env var work
1. [x] fix: Move state (current template, current model element, breadcrumbs) to client as discussed with @Wuestengecko
1. [x] feat: Implement "Click to enlarge" on hover for SVGs which opens a modal dialog with pan/ zoom/ print/ close functionality for a diagram
1. [x] chore: Consider (reuse) plotly license
1. [x] fix: Make print in `diagramViewer.js` work (scaling must be fixed for diagrams with a height which is greater than the width)
1. [x] feat: Show traceback when running into #82
1. [x] chore: Cache headers for statics
1. [x] docs: Update `README.md`/ `CONTRIBUTING.md` 
1. [x] Add some basic tests (and restore the hint in CONTRIBUTING), see #90 
1. [x] Amend commit message subject so that it is marked (`!`) that this PR introduces breaking changes.
1. [x] Wait for new release with version > 0.6.12 of https://github.com/DSD-DBS/py-capellambse which includes https://github.com/DSD-DBS/py-capellambse/pull/510

# Integration in Capella Collaboration Manager

1. [x] Rename the env vars in CCM tool cfg `MODEL_ENTRYPOINT` --> `CME_MODEL` and `ROUTE_PREFIX` --> `CME_ROUTE_PREFIX` Align this with main contributor @MoritzWeber0 of https://github.com/DSD-DBS/capella-collab-manager.git
2. [x] Ensure, that CCM sets env var `CME_LIVE_MODE=0` in the tool config

### CODE REVIEW

**Review by Wuestengecko — COMMENTED** [review]
Here are some initial comments. As the PR is still WIP anyway, I've mostly only looked at the greater ideas, and not too much into the details yet. (Except for the comment about OOB swaps :p )

File: capella_model_explorer/icons.py
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/icons.py:1]: I thought about having `.svg` files instead of py components for SVGs. Can't do that because we depend on Tailwind CSS utility classes which will be listed in the class attribute of the `<svg>` tag sent to the client instead of attrs of the `<svg>` tag itself.

With external `.svg` I would have final values for colors.

Example:

```python
DARK_ACCENT: t.Final[str] = "white"
LIGHT_ACCENT: t.Final[str] = "black"
```

```html
<svg stroke="white">  <!-- dark theme -->
```

```html
<svg stroke="black">  <!-- light theme -->
```

But I need

```html
<svg class="black dark:stroke-white">  <!-- day theme -->
```

File: capella_model_explorer/main.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+try:
+    _1 = state.model
+except AttributeError:
+    state.model = capellambse.loadcli(MODEL_INFO)
+
+state.jinja_env.finalize = _finalize
+state.jinja_env.filters["make_href"] = (
+    lambda obj: obj.name if obj is not None else ""
+)
+state.jinja_env.filters["make_href"] = _make_href_filter
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/main.py:118]: thx, fixed

[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/main.py]: thx, 8ecebe785f616d51785127874a3e1d3647b9a974

File: .github/workflows/build-test-publish.yml [hunk 2]
```diff
@@ -24,7 +24,7 @@ jobs:
           - "3.13"
         include:
           - os: windows-latest
-            python_version: "3.11"
+            python_version: "3.12"
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply .github/workflows/build-test-publish.yml]: reverted

File: capella_model_explorer/components.py [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+def model_information() -> t.Any:
+    """Render the model information including the badge."""
+    return ft.Div(
+        ft.H1(state.model.name, cls="text-xl"),
+        ft.P(
+            f"Capella version: {state.model.info.capella_version}",
+        ),
+        ft.P(
+            "Date created: dummy",
+        ),
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/components.py:116]: yes, is on my radar. hope i will not forget it. i added an item "show metadata of selected model / revision in statusbar, page header or similar" in the opener of this PR

File: capella_model_explorer/main.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+            components.breadcrumbs_refresh_script(),
+            cls=(
+                "h-full",
+                "justify-center",
+                "place-content-center",
+                "place-items-center",
+                "w-full",
+            ),
+            hx_trigger="load",
+            hx_get="/rendered-template",
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/main.py:324]: will do

File: capella_model_explorer/__main__.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+    """Check if a command is available."""
+    try:
+        subprocess.run(
+            cmd,
+            check=True,
+            capture_output=True,
+        )
+    except subprocess.CalledProcessError:
+        print(f"Command '{cmd[0]}' not found. {errmsg}")
+        raise SystemExit(1) from None
```
Comment (Wuestengecko) [inline capella_model_explorer/__main__.py:25]: This error message is very misleading, an arbitrary command can fail for any number of reasons. `npm install` for instance can fail if the network is unavailable, even though npm itself is installed.

**Comment by Wuestengecko** [comment]
> New is also an **always enabled hot template reloading**.

Please add a way to disable that.

uvicorn's live reloading feature injects a websocket connection that reloads the page automatically. Problem with that is that it also tries to reload the page when that websocket connection breaks, for example due to the server going down. This reload then of course fails, because the server is down now, and just shows an error page. So the effect is that you can't just keep an old version of a page opened in the background. I want to be able to do just that, for example to compare against a newer template version, or just to look at the current DOM and make template changes without constantly being thrown around in the page inspector due to the reloads.

Also, the feature should be disabled when deploying the docker image to e.g. the Collab Manager, where the templates are immutable anyway.

File: capella_model_explorer/constants.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+diagramViewer_hash = core.compute_file_hash(diagramViewer_path)
+
+theme_path = "static/js/theme.js"
+theme_hash = core.compute_file_hash(theme_path)
+
+HEADERS: t.Final[list[fh.Link | ft.Script]] = [
+    fh.HighlightJS(langs=["python"]),
+    fh.Link(
+        rel="stylesheet",
+        href=f"{ROUTE_PREFIX}/static/css/main-{cme.__version__}.min.css",
```
Comment (jamilraichouni) [inline capella_model_explorer/constants.py:68]: Also add hash as `?v=HASH` to URL so that the CSS is always reloaded when changed (especially when in `local-dev`, where the app version is stable)

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -57,6 +53,66 @@ We additionally recommend that you set up your editor / IDE as follows.
...
+### Modifying the style of the frontend
+
+Always modify the style via TailwindCSS classes within the Python or Jinja2
+template files.
+
+If not possible (because of collision with TailwindCSS' `prosa` class), modify
+the file `static/css/input.css` directly.
+
+Note, that it is sometimes needed to empty cache and hard reload the app in the
+browser, when styles are not applied correctly.
```
Comment (jamilraichouni) [inline CONTRIBUTING.md:114]: Remove this, when we serve the `main-vx.y.z..min.css` with a `?v=HASH` query parameter.

File: capella_model_explorer/main.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    key_fname="/tmp/.sesskey",
+    live=c.LIVE_MODE,
+    pico=False,
+    routes=ROUTES,
+)
+ar = fh.APIRouter(prefix=c.ROUTE_PREFIX)
+app.static_route_exts(f"{c.ROUTE_PREFIX}/static", "./static")
+
+
+@app.on_event("startup")
```
Comment (jamilraichouni) [inline capella_model_explorer/main.py:51]: Consider

```text
.venv/lib/python3.12/site-packages/starlette/routing.py:865: DeprecationWarning: The `on_event` decorator is deprecated, and will be removed in version 1.0.0. Refer to 
https://www.starlette.io/lifespan/ for recommended approach.
```

see https://www.starlette.io/lifespan/

**Review by Wuestengecko — COMMENTED** [review]
Haven't looked at much of the actual code yet. More important stuff came up, so I don't know when I'll be able to. So here you go.

File: capella_model_explorer/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/main.py]: done.

**Comment by jamilraichouni** [comment]
> > New is also an **always enabled hot template reloading**.
> 
> Please add a way to disable that.
> 
> uvicorn's live reloading feature injects a websocket connection that reloads the page automatically. Problem with that is that it also tries to reload the page when that websocket connection breaks, for example due to the server going down. This reload then of course fails, because the server is down now, and just shows an error page. So the effect is that you can't just keep an old version of a page opened in the background. I want to be able to do just that, for example to compare against a newer template version, or just to look at the current DOM and make template changes without constantly being thrown around in the page inspector due to the reloads.
> 
> Also, the feature should be disabled when deploying the docker image to e.g. the Collab Manager, where the templates are immutable anyway.

One can set `CME_LIVE_MODE=0`. My PR comment was misleading. See also:
```bash
cme run local -h
```

File: pyproject.toml [hunk 9]
```diff
@@ -16,45 +16,47 @@ license = { text = "Apache-2.0" }
...
 classifiers = [
-  "Development Status :: 1 - Planning",
-  "License :: OSI Approved :: Apache Software License",
-  "Natural Language :: English",
-  "Operating System :: OS Independent",
-  "Programming Language :: Python :: 3 :: Only",
-  "Programming Language :: Python :: 3.11",
-  "Programming Language :: Python :: 3.12",
-  "Programming Language :: Python :: 3.13",
+    "Development Status :: 1 - Planning",
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply pyproject.toml:19]: mysterious effect. fixed that. will appear with next push.

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -134,19 +137,28 @@ repos:
...
+        # depends on https://github.com/remarkjs/remark/tree/main/packages/remark-parse
+        name: prettier-md
         language: system
-        entry: sh -c 'make frontend/node_modules && ./frontend/node_modules/.bin/eslint "$@"' --
+        entry: sh -c 'node_modules/.bin/prettier --write --parser markdown "$@"' --
         require_serial: true
-        args: ['--fix', '--config', 'frontend/eslint.config.js']
-        types: []
-        files: '^frontend/.*\.(js|jsx)$'
+        files: '\.md$'
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply .pre-commit-config.yaml:164]: fixed that. will appear with next push.

File: capella_model_explorer/components.py [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+    )
+
+
+def breadcrumbs_refresh_script() -> t.Any:
+    return ft.Script(
+        "htmx.ajax("
+        "'GET', '/breadcrumbs', {swap: 'outerHTML', target:'#breadcrumbs'}"
+        ");"
+        "console.log('breadcrumbs refreshed');"
+    )
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/components.py]: done. will appear with next push.

File: capella_model_explorer/components.py [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+
+
+def model_object_button(model_element: dict) -> t.Any:
+    template = t.cast(templates.Template, state.template)
+    return ft.Div(
+        ft.Div(model_element["name"]),
+        ft.Div(model_element["uuid"], cls="text-xs text-sky-700")
+        if state.show_uuids
+        else None,
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/components.py:167]: see https://github.com/DSD-DBS/capella-model-explorer/discussions/102

File: pyproject.toml [hunk 13]
```diff
@@ -75,7 +77,7 @@ python_version = "3.11"
 
 [[tool.mypy.overrides]]
 # Untyped third party libraries
-module = ["diff_match_patch"]
+module = ["diff_match_patch", "fasthtml.*"]
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply pyproject.toml]: thx. done. will disappear with next push.

File: static/README.md
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply static/README.md:1]: idea was to document how the `.ico` can be made. forgot to push the locally got ignored `dev` subfolder.
after we checked that chrome understands to display `.svg` favicons directl too, i remove this anyway.

File: capella_model_explorer/__main__.py [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+        )
+    else:
+        raise SystemExit(f"Templates directory '{c.TEMPLATES_DIR}' not found.")
+    cmd.append(c.DOCKER_IMAGE_NAME)
+    print(" ".join(cmd))
+    subprocess.run(
+        cmd,
+        check=True,
+        capture_output=False,
+    )
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/__main__.py:91]: thx. done.

File: Dockerfile [hunk 15]
```diff
@@ -1,61 +1,84 @@
...
+RUN uv venv && \
+  # install app
+  uv pip install --no-cache-dir . && \
+  # Install elk.js automatically into a persistent local cache directory
+  # in order to prepare the elk.js execution environment ahead of time.
+  uv run python3 -c "from capellambse_context_diagrams import install_elk; install_elk()"
+# }}}
+
+# install Nodep kgs and build the CSS {{{
+RUN npm clean-install && uv run python3 -m capella_model_explorer build css
```
Comment (Wuestengecko) [inline Dockerfile:68]: We don't even need a multi-stage build, we can just remove the `node_modules` right here. They're not needed after this step anymore, as far as I can tell.

```suggestion
RUN npm clean-install && uv run python3 -m capella_model_explorer build css && rm -rf node_modules
```

File: .prettierrc
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply .prettierrc:1]: You forgot to sort the entries alphabetically 😇

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -57,6 +53,66 @@ We additionally recommend that you set up your editor / IDE as follows.
...
+The app will be available at the location printed to the terminal when
+the console confirms `Application startup complete.`.
+
+### Modifying the style of the frontend
+
+Always modify the style via TailwindCSS classes within the Python or Jinja2
+template files.
+
+If not possible (because of collision with TailwindCSS' `prosa` class), modify
+the file `static/css/input.css` directly.
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply CONTRIBUTING.md:111]: I know. At the same time it is possible and people may:

- start to make use of it
- know, that this is even possible.

File: Dockerfile [hunk 17]
```diff
@@ -1,61 +1,84 @@
...
+# }}}
+
+# clean up as root {{{
+USER root
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply Dockerfile:72]: Thanks, I replace `0:0` because it is more readable.

File: entrypoint.sh [hunk 18]
```diff
@@ -3,10 +3,5 @@
...
 
-sed -i "s|__ROUTE_PREFIX__|${ROUTE_PREFIX}|g" ./frontend/dist/static/env.js
-sed -i "s|href=\"/|href=\"${ROUTE_PREFIX}/|g" ./frontend/dist/index.html
-sed -i "s|src=\"/|src=\"${ROUTE_PREFIX}/|g" ./frontend/dist/index.html
-
-export MODEL="${MODEL_ENTRYPOINT}"
-export TEMPLATES_DIR="/views"
-exec uvicorn --host=0.0.0.0 --port=8000 capella_model_explorer.backend.main:app
+export CME_TEMPLATES_DIR="/app/templates"
+uv run -m capella_model_explorer run local
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply entrypoint.sh:7]: Tried that and it did not work. I modified the entrypoint. It activates the venv and runs the code now.
uv is not directly invoked anymore.

File: README.md [hunk 19]
```diff
@@ -12,7 +12,6 @@ A webapp for exploring Capella models through simple "auto-generated" textual
 and graphical views.
 
 **Longer story**:
-
 We see a larger non-MBSE crowd struggling with the things hidden in the model.
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply README.md:15]: I keep the empty line as it was before.

File: templates/common_macros.html.j2 [hunk 20]
```diff
@@ -100,7 +100,7 @@
               <td style="word-break:normal">{{ attr }}</td>
               <td>
                 {%- if value.as_svg -%}
-                  {{ value.as_svg | safe }}
+                  {{ value }}
```
Comment (jamilraichouni) [inline templates/common_macros.html.j2:103]: @Wuestengecko this breaks a template.
Try

```shell
CME_MODEL=git+https://github.com/DSD-DBS/Capella-IFE-sample.git
```
and then go to:

System Interfaces. There is just one named `I_IFE_Crew`.

Comment (jamilraichouni) [reply templates/common_macros.html.j2:103]: Question is, how we get the "Click to enlarge" feature, when we cannot apply above modification.

File: capella_model_explorer/__main__.py [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+    input_css = pathlib.Path("static/css/input.css")
+    if not input_css.is_file():
+        raise SystemExit(f"Input CSS file not found: {input_css}")
+    tailwind_cmd = [
+        exe,
+        "--minify",
+        "--input",
+        str(input_css),
+        "--output",
+        f"static/css/main-{capella_model_explorer.__version__}.min.css",
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/__main__.py:182]: So far the current implementation just works fine. The last time that I had to empty the browser's cache and do a hard reload was when the `main.min.css` source URL had no `?v=<HASH>` query parameter.

File: templates/20-System Analysis/system-actor.html.j2 [hunk 22]
```diff
@@ -35,7 +35,7 @@
 
 <h2>Context</h2>
 <p>The context of {{ object.name }} is summarized by the diagram below:</p>
-{{ object.context_diagram.render("svg", display_derived_interfaces=True) | safe }}
+{{ object.context_diagram.render(None, display_derived_interfaces=True) | safe }}
```
Comment (jamilraichouni) [inline templates/20-System Analysis/system-actor.html.j2:38]: As mentioned at other locations. This change breaks the rendering of the diagram and we need to talk about that.

File: .github/workflows/build-test-publish.yml [hunk 23]
```diff
@@ -74,7 +76,7 @@ jobs:
       - name: Setup Python
         uses: actions/setup-python@v5
         with:
-          python-version: "3.11"
+          python-version: "3.12"
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply .github/workflows/build-test-publish.yml:79]: IMO and as mentioned earlier I recommend to test against 3.11 too. I adapt the code so that 3.11, 3.12, and 3.13 are supported. Can you cross-check the workflow after my next push, please?

File: capella_model_explorer/__main__.py [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+            + templ_reload_hlp
+            + opt_env_vars_hlp
+            + " CME_DOCKER_IMAGE_NAME"
+            f" (default: {c.Defaults.docker_image_name})"
+            f",{model_hlp}"
+            f",{templates_dir_hlp}"
+            f",{route_prefix_hlp}"
+            f",{live_hlp}"
+            f",{port_hlp}"
+        ),
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/__main__.py:274]: Yes, good proposal! I want to go back to `click` and [serve](https://docs.fastht.ml/api/core.html#serve) instead of subprocess runs of `uvicorn` anyway.
Can we postpone and keep that for a next PR?

File: pyproject.toml [hunk 25]
```diff
@@ -26,23 +26,34 @@ classifiers = [
...
+test = [
+  "coverage",
+  "psutil",
+  "pytest",
+  "pytest-cov",
+  "pytest-xdist",   # to be able to run soecific tests sequentially
+  "requests",
+  "types-psutil",
+  "types-requests",
+]
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply pyproject.toml:53]: Leftovers. I removed some of them and kept:

```toml
test = [
  "coverage",
  "pytest",
  "pytest-cov",
]
```

File: capella_model_explorer/state.py [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+
+if t.TYPE_CHECKING:
+    import capellambse
+
+JINJA_ENV: jinja2.Environment
+
+
+class Breadcrumb(pydantic.BaseModel):
+    label: str
+    url: str
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/state.py:23]: The module `types.py` has been removed since its content has been moved into `reports.py`. The `Breadcrumb` helper class has been moved to `components.py` since this is the only module using it.

File: capella_model_explorer/components.py [hunk 27]
```diff
@@ -0,0 +1,623 @@
...
+            "max-h-12",
+            "px-4",
+            "py-0",
+            "print:hidden",
+            "sticky",
+            "top-0",
+            "w-full",
+            # ensure that SVGs in rendered reports won't overlap in
+            # dark mode due to CSS filters
+            "z-50",
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/components.py:271]: Well, when I added this comment we had a quite weird issue only in dark mode and in combination with our special SVG filters. That was special.
This is why I left a comment here. Somehow this behavior changed (maybe due to the move to TailwindCSS v4.0).
I remove the comment now.

File: capella_model_explorer/app.py [hunk 28]
```diff
@@ -0,0 +1,311 @@
...
+    state.model = capellambse.loadcli(c.MODEL)
+    state.models = {core.model_id(state.model): state.model}
+    logger.info("Model load complete.")
+    logger.info(
+        "Waiting for templates to load from directory '%s'", c.TEMPLATES_DIR
+    )
+    reports.load_templates()
+    state.JINJA_ENV = jinja2.Environment(
+        loader=jinja2.FileSystemLoader(c.TEMPLATES_DIR)
+    )
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/app.py]: Adding `autoescape=True` does not solve https://github.com/DSD-DBS/capella-model-explorer/pull/81#discussion_r1956489677

[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capella_model_explorer/app.py]: That seems to have been an unrelated problem, see DSD-DBS/py-capellambse#510.

File: .github/workflows/build-test-publish.yml [hunk 29]
```diff
@@ -14,44 +14,47 @@ jobs:
...
       matrix:
-        os: [ubuntu-latest]
+        os: [ubuntu-latest, windows-latest]
         python_version:
           - "3.11"
           - "3.12"
           - "3.13"
-        include:
-          - os: windows-latest
-            python_version: "3.11"
```
Comment (Wuestengecko) [inline .github/workflows/build-test-publish.yml:27]: I don't think we need to run Windows tests for all Python versions, the minimum supported one should be enough. The other Python versions are already covered by the Linux tests.

Reasons being that a) Windows tests usually run a little slower compared to Linux, and b) they consume twice the CI time on Github (though this is a public repo, so we don't actually have to pay that).

File: capella_model_explorer/__main__.py [hunk 30]
```diff
@@ -0,0 +1,351 @@
...
+        print(shlex.join(tailwind_cmd))
+        return subprocess.Popen(tailwind_cmd)
+    print(shlex.join(tailwind_cmd))
+    subprocess.check_call(tailwind_cmd)
+    return None
+
+
+def build_image():
+    if pathlib.Path.cwd().name != "capella-model-explorer":
+        raise SystemExit("Run this command from the root of the project.")
```
Comment (Wuestengecko) [inline capella_model_explorer/__main__.py:198]: This condition makes an assumption about the _outside_ of the project, which we have no influence over. This breaks for example if you have a directory structure like `.../capella/model-explorer`.

If you want to check for the repo root, check if `.git/` and `capella_model_explorer/` are present instead.

File: capella_model_explorer/__main__.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capella_model_explorer/__main__.py:185]: So basically, since we have `?v=` on main.css as well now, we don't need to do the versioned filename crap anymore. Let's remove that then :)

File: capella_model_explorer/__main__.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capella_model_explorer/__main__.py:280]: > Can we postpone and keep that for a next PR, please?

That part I don't agree with. We don't have to wait until click to get a nice CLI, and this also has nothing to do with serve() vs. a subprocess. And there's really not a lot to do anyway - can you please cherry-pick f5bf391725126c754eed5612391ab59c17d240ce?

File: .github/workflows/build-test-publish.yml [hunk 31]
```diff
@@ -14,44 +14,47 @@ jobs:
...
   pre-commit:
+    name: Pre-Commit with Python ${{matrix.python_version}}
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python_version:
+          - "3.11"
+          - "3.12"
+          - "3.13"
```
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply .github/workflows/build-test-publish.yml:50]: Idea is to be on the very safe side. At the same time I see that it makes a difference of approx. 3 mins per CI run. That can annoy and I revert this to run for the minimum supported version only for now.

File: capella_model_explorer/__main__.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/__main__.py:185]: Yeah, you're right. Germans say "Doppelt hält besser" here. But no need for "doppelt" here.

File: capella_model_explorer/state.py [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+
+    from capella_model_explorer import reports
+
+jinja_env: jinja2.Environment
+
+
+idle_time_gauge = prometheus_client.Gauge(
+    "idle_time_minutes",
+    "Time in minutes since the last user interaction",
+)
```
Comment (Wuestengecko) [inline capella_model_explorer/state.py:23]: Note this rename (`idletime_minutes` => `idle_time_minutes`) is a breaking change for monitoring setups. We need to make sure to update that and keep it consistent everywhere after the merge. That's why I proposed to rename the test functions rather than the metric in my earlier comment.

File: capella_model_explorer/__main__.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (jamilraichouni) [reply capella_model_explorer/__main__.py:280]: Thank you! I cherry-picked and fixed a bit.

The command `run` had no summary line. Maybe this was intended.
The default for the Docker image name was not documented in the help  anymore and
`cme run -h` referred to some not-existing subcommand target `cme build container`.

File: capella_model_explorer/state.py (hunk 32 above)
Comment (jamilraichouni) [reply capella_model_explorer/state.py:23]: Yes, mentioned that in the commit message already and your message led to another todo item in the PR description. Thanks!

File: templates/20-System Analysis/system-actor.html.j2 (hunk 22 above)
Comment (jamilraichouni) [reply templates/20-System Analysis/system-actor.html.j2]: My fault. Forgot to remove the ` | safe` part.

File: templates/common_macros.html.j2 (hunk 20 above)
Comment (jamilraichouni) [reply templates/common_macros.html.j2:103]: An update of capellambse (see https://github.com/DSD-DBS/py-capellambse/pull/510) will fix this.

**Comment by Wuestengecko** [comment]
Found a few more, very minor issues, which I've gone ahead and quickly fixed myself. With that, I'm happy for now. LGTM.
